### PR TITLE
Fix componentDidUpdate when updating by setState on v16

### PR DESCRIPTION
**Pull request by koba04**
In v16, ShallowRenderer doesn't call `componentDidUpdate` even updating by `setState`. 
https://github.com/facebook/react/pull/10372
This PR is to fix this.

### CODE REVIEW

**Comment by koba04** [comment]
l know ShallowRenderer v16 calls cDU without a context argument but I didn't include that because I think it should be fixed separately.

File: packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js [hunk 1]
```diff
@@ -143,6 +143,7 @@ class ReactThirteenAdapter extends EnzymeAdapter {
     let isDOM = false;
     let cachedNode = null;
     return {
+      disableComponentDidUpdateOnSetState: true,
```
Comment (ljharb) [inline packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: This seems like a really awkward way to extend the adapter interface.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx [hunk 2]
```diff
@@ -3403,8 +3403,7 @@ describe('shallow', () => {
 
     context('updating state', () => {
       // NOTE: There is a bug in react 16 shallow renderer where prevContext is not passed
-      // into componentDidUpdate. Skip this test for react 16 only. add back in if it gets fixed.
```
Comment (ljharb) [inline packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: prevContext *should* be passed into componentDidUpdate; is react not going to fix this?

File: packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js (hunk 1 above)
Comment (koba04) [reply packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: What about this?

```js
return {
  options: {
    disableComponentDidUpdateOnSetState: true
  }
```

I can do this without to extend the adapter interface but it requires to access the React version directly from ShallowWrapper. I think ShallowWrapper should depend on adapters only.

File: packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx (hunk 2 above)
Comment (koba04) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Yes. #1139 is a PR for that.
You can see the change at [Unreleased] section in CHANGELOG.md 

https://github.com/facebook/react/blob/master/CHANGELOG.md

Comment (ljharb) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Then if that's fixed, why do we need an option in the adapter?

Comment (koba04) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: The PR is to support that componentDidUpdate no longer receives prevContext argument, which is not only ShallowRenderer.
This PR is to support that ShallowRenderer no longer call componentDidUpdate.
These are different fixes so I added the options each other.

Comment (ljharb) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Why does componentDidUpdate not receive a prevContext argument anymore? It seems like it should.

Comment (koba04) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: The reason seems to be the following. cc: @bvaughn

> Even with that parameter present- if context isn't managed in the same way as props and state (PR #8655) then it could be inconsistent between the instance and the fiber.

> Update: Chatted in meatspace with @acdlite about this. He pointed out another existing stack limitation with fiber WRT shouldComponentUpdate. (Seems like context is already flaky but fiber will make it more so.) The only safe way to use context is to pass singleton/static props (eg the way react-redux Provider passes down store) that components subscribe to for changes.

https://github.com/facebook/react/pull/8631#issuecomment-270450606

Comment (ljharb) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: Wow, that seems like a pretty major change to the way context works then.

I've not found context to be flaky at all in React <= 15.

@lelandrichardson any suggestions for a better name for this part of the adapter interface?

Comment (bvaughn) [reply packages/enzyme-test-suite/test/ShallowWrapper-spec.jsx]: > Why does componentDidUpdate not receive a prevContext argument anymore? It seems like it should.

@ljharb Check out facebook/react/pull/8631 for a description of the options we considered.

> The only safe way to use context is to pass singleton/static props...that components subscribe to for changes.

Yes. That sounds accurate.

File: packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js (hunk 1 above)
Comment (lelandrichardson) [reply packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: I wonder if the option should actually be the opposite. `enableComponentDidUpdateOnSetState`, making it an opt-in flag for adapters, and one that only react 16 will be using?

Comment (koba04) [reply packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: @lelandrichardson I think ShallowRenderer won't support componentDidUpdate in future releases so the flag is required by only versions before v16. Using `enableComponentDidUpdateOnSetState` means that enzyme has to set the flag for future versions of React permanently. So I used `disableComponentDidUpdateOnSetState` for this.
But it might make sense to use `enableComponentDidUpdateOnSetState` and set the flag until dropping v15 support.

Comment (lelandrichardson) [reply packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: that makes sense.  I think there might be more use cases for adding "options" for the adapter. I think i like giving them their own namespace though: `options.enableComponentDidUpdateOnSetState` or something.

Also, I wonder if it's better to put this on the adapter instance itself instead of the object it returns from `createRenderer`?  This way, we can set an empty `this.options = { ... }` in the `EnzymeAdapter` base constructor, which would mean people wouldn't have to set defaults.

Thoughts?

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js [hunk 3]
```diff
@@ -145,6 +145,12 @@ function nodeToHostNode(_node) {
 }
 
 class ReactSixteenAdapter extends EnzymeAdapter {
+  constructor() {
+    super();
+    this.options = { ...this.optins,
```
Comment (lelandrichardson) [inline packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: typo

File: packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js (hunk 1 above)
Comment (koba04) [reply packages/enzyme-adapter-react-13/src/ReactThirteenAdapter.js]: That makes sense! I've updated this PR.

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js (hunk 3 above)
Comment (koba04) [reply packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: Fixed

File: packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js [hunk 4]
```diff
@@ -145,6 +145,12 @@ function nodeToHostNode(_node) {
 }
 
 class ReactSixteenAdapter extends EnzymeAdapter {
+  constructor() {
+    super();
+    this.options = { ...this.options,
```
Comment (ljharb) [inline packages/enzyme-adapter-react-16/src/ReactSixteenAdapter.js]: `...this.options` needs to go on its own line

File: packages/enzyme-test-suite/package.json [hunk 5]
```diff
@@ -40,4 +40,4 @@
     "react": "^15.5.0",
     "react-dom": "^15.5.0"
   }
-}
+}
```
Comment (ljharb) [inline packages/enzyme-test-suite/package.json]: all files need to have a trailing newline

File: packages/enzyme/src/ShallowWrapper.js [hunk 6]
```diff
@@ -369,7 +369,39 @@ class ShallowWrapper {
...
     this.single('setState', () => {
       withSetStateAllowed(() => {
-        this.instance().setState(state, callback);
+        const adapter = getAdapter(this[OPTIONS]);
+        const instance = this.instance();
+        const prevProps = instance.props;
+        const prevState = instance.state;
+        const prevContext = instance.context;
+        let shouldRender = true;
+        // dirty hack: avoid calling shouldComponentUpdate twice
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: why is this needed? can you elaborate in the comment?

Comment (koba04) [reply packages/enzyme/src/ShallowWrapper.js]: Added a comment. Does this make sense?

File: packages/enzyme/src/ShallowWrapper.js [hunk 7]
```diff
@@ -369,7 +369,42 @@ class ShallowWrapper {
...
+        let originalShouldComponentUpdate;
+        if (
+          this[OPTIONS].lifecycleExperimental &&
+          adapter.options.enableComponentDidUpdateOnSetState &&
+          instance &&
+          typeof instance.shouldComponentUpdate === 'function'
+        ) {
+          originalShouldComponentUpdate = instance.shouldComponentUpdate;
+          instance.shouldComponentUpdate = (...args) => {
+            shouldRender = originalShouldComponentUpdate.apply(instance, args);
```
Comment (ljharb) [inline packages/enzyme/src/ShallowWrapper.js]: should this function also `instance.shouldComponentUpdate = originalShouldComponentUpdate`?

I wonder if this should use `sinon` as a dependency rather than overwriting; what about things like enumerability, setters, etc?

**Comment by lelandrichardson** [comment]
made small tweak based on @ljharb's comment. merging now in interest of time. Thanks!

File: packages/enzyme/src/ShallowWrapper.js (hunk 7 above)
Comment (koba04) [reply packages/enzyme/src/ShallowWrapper.js]: > I wonder if this should use sinon as a dependency rather than overwriting; what about things like enumerability, setters, etc?

Yeah, I can write this like the following if I use `sinon`.

```js
        let shouldRender = true;
        let spy;
        if (instance.shouldComponentUpdate) {
          spy = sinon.spy(instance, 'shouldComponentUpdate');
        }
        instance.setState(state, callback);
        if (spy) {
          shouldRender = spy.getCall(0).returnValue;
          spy.restore();
        }
```

Also `sinon` seems to have a better support about enumerability so I think it's worth to consider this after releasing v3. ES2015 Proxy might be an option for that.

**Comment by koba04** [comment]
Thanks!

File: packages/enzyme/src/ShallowWrapper.js (hunk 7 above)
Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: I'd prefer to avoid Proxy; I'm not sure it would work here anyways, and it would very tightly constrict the engines enzyme can run on.

Comment (koba04) [reply packages/enzyme/src/ShallowWrapper.js]: True. To support `< Node v6`, `sinon.spy` would be nice.

Comment (ljharb) [reply packages/enzyme/src/ShallowWrapper.js]: A followup PR is always appreciated :-D

Comment (koba04) [reply packages/enzyme/src/ShallowWrapper.js]: I'll work on it 👍

**Comment by austinh** [comment]
Has this been merged into v3.1.0? Because I cannot get componentDidUpdate to run when using setState

**Comment by ljharb** [comment]
Yes, it's in v3.1; please file a new issue if you have trouble still.
